src/task/builder.rs

```diff
@@ -151,7 +157,36 @@ impl Builder {
         });
 
         // Run the future as a task.
-        unsafe { TaskLocalsWrapper::set_current(&wrapped.tag, || smol::run(wrapped)) }
+        IS_CURRENT_BLOCKING.with(|is_current_blocking| {
+            let count = is_current_blocking.get();
+            let res = if count == 0 {
+                // increase the count
+                is_current_blocking.replace(1);
+
+                // The first call should use run.
+                unsafe {
+                    TaskLocalsWrapper::set_current(&wrapped.tag, || {
+                        let res = smol::run(wrapped);
+                        is_current_blocking.replace(is_current_blocking.get() - 1);
+                        res
+                    })
+                }
```

Comment:
This code is duplicated with the branch below. It seems the only difference is between:
```rust
is_current_blocking.replace(1);
```
and
```rust
is_current_blocking.replace(is_current_blocking.get() + 1);
```
We should probably only branch on that and share the `TaskLocalsWrapper` logic.

Reply:
fixec